### PR TITLE
Fix Scoping Issues In Naive Codegen

### LINKED ISSUE #1065: Scoping Issues in ICO Naive Backend
There are **two** very related bugs in the ICO naive backend

First: There is a bug in the cxx-naive-ico code generator in conjunction with an if condition. Each reduction is supposed to span its own scope, to encapsulate the sparse index iterator from the outside world. This fails in the case of an if condition, where a scope over the whole if condition is spanned. This causes problems if both branches or even the conditional contain a reduction or loop fill statement, because the linear / sparse index will not be re-set between the individual iterations. A close to minimal example follows:

**dusk code**
```
@stencil
def if_condition(f: Field[Edge], nx: Field[Edge], L: Field[Edge], A: Field[Cell], edge_orientation: Field[Cell > Edge], 
                 f_x: Field[Cell]):
    with levels_downward:
        if sum_over(Cell > Edge, 1) < 4:
            f_x = sum_over(Cell > Edge, L)
        else:
            f_x = sum_over(Cell > Edge, L)
```

**generated code**
```
for (auto const &loc : getCells(LibTag{}, m_mesh)) {
  {
    int sparse_dimension_idx0 = 0;
    if ((reduce(LibTag{}, m_mesh, loc, (::dawn::float_type)0,
                std::vector<::dawn::LocationType>{
                    ::dawn::LocationType::Cells,
                    ::dawn::LocationType::Edges},
                [&](auto &lhs, auto red_loc1) {
                  lhs += (int)1;
                  sparse_dimension_idx0++;
                  return lhs;
                }) < (int)4)) {
      m_f_x(deref(LibTag{}, loc)) =
          reduce(LibTag{}, m_mesh, loc, (::dawn::float_type)0,
                  std::vector<::dawn::LocationType>{
                      ::dawn::LocationType::Cells,
                      ::dawn::LocationType::Edges},
                  [&](auto &lhs, auto red_loc1) {
                    lhs += m_L(deref(LibTag{}, red_loc1));
                    sparse_dimension_idx0++;
                    return lhs;
                  });
    } else {
      m_f_x(deref(LibTag{}, loc)) =
          reduce(LibTag{}, m_mesh, loc, (::dawn::float_type)0,
                  std::vector<::dawn::LocationType>{
                      ::dawn::LocationType::Cells,
                      ::dawn::LocationType::Edges},
                  [&](auto &lhs, auto red_loc1) {
                    lhs += m_L(deref(LibTag{}, red_loc1));
                    sparse_dimension_idx0++;
                    return lhs;
                  });
    }
  }
```

Second: A very related bug occurs if a local field is demoted to a scalar, written to in one reduction and then consumed in a subsequent reduction. Due to the isolation into different scopes the later reduction can't access the emitted scalar.  

**dusk code**
```
@stencil
def sph(xn: Field[Vertex], yn: Field[Vertex], xc: Field[Cell], yc: Field[Cell], 
        fn: Field[Vertex], f_intp: Field[Cell], h: Field[Cell], pi: Field[Cell]):
    wij: Field[Cell > Vertex > Edge > Vertex]
    qij: Field[Cell > Vertex > Edge > Vertex]
    Wc: Field[Cell]

    with levels_upward:
        with sparse[Cell > Vertex > Edge > Vertex]:
            qij = sqrt((xc-xn)*(xc-xn) + (yc-yn)*(yc-yn))/h
            wij = 1./(pi*h*h)*exp(-qij*qij)

        Wc = sum_over(Cell > Vertex > Edge > Vertex, wij)    
        f_intp = sum_over(Cell > Vertex > Edge > Vertex, 1/Wc*wij*fn)
```
**generated code**
```
for(auto const& loc : getCells(LibTag{}, m_mesh)) {
  {
    int sparse_dimension_idx0 = 0;
    ::dawn::float_type __local_Wc_2_224 =
        reduce(LibTag{}, m_mesh, loc, (::dawn::float_type)0,
                std::vector<::dawn::LocationType>{
                    ::dawn::LocationType::Cells, ::dawn::LocationType::Vertices,
                    ::dawn::LocationType::Edges, ::dawn::LocationType::Vertices},
                [&](auto& lhs, auto red_loc1) {
                  lhs += m_wij(deref(LibTag{}, loc), sparse_dimension_idx0);
                  sparse_dimension_idx0++;
                  return lhs;
                });
  }
  {
    int sparse_dimension_idx0 = 0;
    m_f_intp(deref(LibTag{}, loc)) =
        reduce(LibTag{}, m_mesh, loc, (::dawn::float_type)0,
                std::vector<::dawn::LocationType>{
                    ::dawn::LocationType::Cells, ::dawn::LocationType::Vertices,
                    ::dawn::LocationType::Edges, ::dawn::LocationType::Vertices},
                [&](auto& lhs, auto red_loc1) {
                  lhs += ((((int)1 / __local_Wc_2_224) *
                          m_wij(deref(LibTag{}, loc), sparse_dimension_idx0)) *
                          m_fn(deref(LibTag{}, red_loc1)));
                  sparse_dimension_idx0++;
                  return lhs;
                });
  }
}
```

### PR DESCRIPTION
**Pull request by mroethlin**
## Technical Description

In the naive backend, each reduction was wrapped into its own scope. This caused some problems in if statements where a reduction was part of the condition, as well as situations where a temporary field was demoted to a local variable. This PR addresses this by removing the scope. The local linear index is now introduced in the lambda capture, which is made possible by marking the lambda as `mutable`. 

### Resolves / Enhances

Fixes #1065 

### Example

A real life example could be
```
  with levels_downward:
    if sum_over(Cell > Edge, 1) < 4:
        f_x = ... #do something at the boundary
    else:
        f_x = sum_over(Cell > Edge, L)
```

### Testing

A test to the effect of the above was added to the unstructured integration checking. 

### Dependencies

This PR is independent

### CODE REVIEW

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

File: dawn/src/dawn/CodeGen/CXXNaive-ico/ASTStencilBody.cpp [hunk 1]
```diff
@@ -371,18 +371,17 @@ void ASTStencilBody::visit(const std::shared_ptr<iir::ReductionOverNeighborExpr>
...
-    ss_ << ", [&](auto& lhs, auto red_loc" << (reductionDepth_ + 1) << ") { ";
-    // generate this next red_loc only if the rhs contains further reductions
-    FindReduceOverNeighborExpr redFinder;
-    expr->getRhs()->accept(redFinder);
-    if(redFinder.hasReduceOverNeighborExpr()) {
-      ss_ << "int " << ASTStencilBody::ReductionSparseIndexVarName(reductionDepth_ + 1) << " = 0;";
-    }
+    ss_ << ", [&, " + ASTStencilBody::ReductionSparseIndexVarName(reductionDepth_) +
+               " = int(0)](auto& lhs, auto red_loc"
+        << (reductionDepth_ + 1) << ") mutable { ";
```
Comment (Stagno) [inline dawn/src/dawn/CodeGen/CXXNaive-ico/ASTStencilBody.cpp:384]:
```suggestion
               " = int(0)](auto& lhs, auto " 
       << ASTStencilBody::ReductionIndexVarName(reductionDepth_ + 1)
       << ") mutable { ";
```

File: dawn/test/integration-test/unstructured/CMakeLists.txt [hunk 2]
```diff
@@ -21,25 +21,38 @@ set(generated_stencil_codes generated_accumulateEdgeToCell.hpp
...
   generated_sparseAssignment1.hpp
+  generated_SparseAssignment1.hpp
   generated_sparseAssignment2.hpp
+  generated_SparseAssignment2.hpp
   generated_sparseAssignment3.hpp
+  generated_SparseAssignment3.hpp
   generated_sparseAssignment4.hpp
+  generated_SparseAssignment4.hpp
   generated_sparseAssignment5.hpp
+  generated_SparseAssignment5.hpp
```
Comment (Stagno) [inline dawn/test/integration-test/unstructured/CMakeLists.txt:47]: duplicates

**Review by Stagno — CHANGES_REQUESTED** [review]
Tiny changes. Overall good to go.

**Comment by mroethlin** [comment]
launch jenkins

**Review by Stagno — APPROVED** [review]

**Comment by mroethlin** [comment]
launch jenkins

**Pull request merged**